2024-06-20-investigating-code-quality-from-pr-data.md

```diff
@@ -0,0 +1,75 @@
+
+The role of a software developer is multi-faceted, and writing code is just one part of the job. Therefore, if GenAI coding tools allow us to write code more quickly, but that code ultimately takes longer to review or is harder to maintain, is it really making us more productive?
+
+
+### So, what gives us an indication of code quality?
+
+When a developer wants to make changes to a code base, they raise a pull request (PR) which contains the proposed changes to the code and a written summary of the changes made. Other developers will then review this PR, leaving comments or suggestions, before ultimately deciding whether to approve the changes.
+
+PRs contain a huge amount of valuable data that can start to give us an idea of the code's quality. If the [CI/CD pipeline](https://about.gitlab.com/topics/ci-cd/) repeatedly fails, the code may not have been thoroughly tested or properly formatted. If a PR receives lots of review comments, it's likely that there are a number of changes that need to be made. Equally, if it takes weeks for a PR to be approved, it may be a reflection of the amount of work required to get the code to a point that the other developers are happy with it (either that, or the developers are just slow to submit their reviews...)
+
+### Enter LLMs
+
+There are some common "pain points" with AI-generated code. Things like not adhering to project conventions, not using functions that exist in other parts of the code base, producing algorithms with sub-optimal performance, or code that is hard to read, is often a clue that code may have been AI-generated, and are likely to be picked up in a review. 
+
+Therefore, what's written in comments is also a valuable source of information. For example, what do reviewers frequently suggest needs to be changed? Is a developer frustrated to make this suggestion (perhaps they've made the same suggestion several times already)? Are developers generally polite to their colleagues, but harsher on code that they suspect is AI-generated?
+
+These are all questions we can begin to answer with the help of LLMs (Large Language Models), which are able to interpret and generate human language. Therefore, we can pass review comments, or even pieces of code, to an LLM and use them to start to answer these questions.
+
+Interestingly, this also gives us an insight into the dynamics within development teams, but more on that later. 
+
+### Pulling this together into a tool
+
+We provide our analyser with a GitHub or GitLab URL, and it then scrapes the PRs that have been raised within that repository. We then take the data and process it, to produce a range of different metrics, which are we have grouped into deterministic or AI. Our deterministic metrics use data such as the number of files changed, number of comments, requests for change, PR duration and PR contributors. The AI metrics use LLMs to analyse things like comment tone and subject, and detect disagreements within comment threads.
```

Comment:
*We then take the data and process it, to produce a range of different metrics which we have grouped into deterministic or AI.
*requests for change, PR duration, and PR contributors.